File: backend/app/api/routes/modules.py
from fastapi import APIRouter, HTTPException, Query, Depends
from pydantic import BaseModel
from typing import Optional, Dict, Any

from app.core.msf_client import msf_client
from app.api.routes.auth import get_current_active_user

router = APIRouter()


class ModuleExecuteRequest(BaseModel):
    options: Dict[str, Any] = {}
    payload: Optional[str] = None
    payload_options: Optional[Dict[str, Any]] = None


@router.get("/types")
async def list_module_types(user=Depends(get_current_active_user)):
    """List available module types."""
    return {
        "types": [
            {"id": "exploit", "name": "Exploits", "description": "Exploitation modules"},
            {"id": "payload", "name": "Payloads", "description": "Payload modules"},
            {"id": "auxiliary", "name": "Auxiliary", "description": "Auxiliary modules (scanners, fuzzers, etc.)"},
            {"id": "post", "name": "Post", "description": "Post-exploitation modules"},
            {"id": "encoder", "name": "Encoders", "description": "Payload encoders"},
            {"id": "nop", "name": "NOPs", "description": "NOP generators"},
        ]
    }


@router.get("/stats")
async def get_module_stats():
    """Get module count statistics (public endpoint for health checks)."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        stats = await msf_client.get_stats()
        return stats
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/search")
async def search_modules(
    q: str = Query(..., min_length=2, description="Search query"),
    type: Optional[str] = Query(None, description="Module type filter"),
    user=Depends(get_current_active_user)
):
    """Search for modules."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        results = await msf_client.search_modules(q, type)
        return {"results": results, "count": len(results)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{module_type}")
async def list_modules(
    module_type: str,
    offset: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    search: Optional[str] = None,
    user=Depends(get_current_active_user)
):
    """List modules of a specific type with pagination."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    valid_types = ["exploit", "payload", "auxiliary", "post", "encoder", "nop"]
    if module_type not in valid_types:
        raise HTTPException(status_code=400, detail=f"Invalid module type. Must be one of: {valid_types}")

    try:
        modules = await msf_client.list_modules(module_type)

        # Filter by search query if provided
        if search:
            search_lower = search.lower()
            modules = [m for m in modules if search_lower in m.lower()]

        total = len(modules)
        modules = modules[offset:offset + limit]

        return {
            "type": module_type,
            "modules": modules,
            "total": total,
            "offset": offset,
            "limit": limit
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{module_type}/{module_name:path}/info")
async def get_module_info(module_type: str, module_name: str, user=Depends(get_current_active_user)):
    """Get detailed information about a specific module."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        info = await msf_client.get_module_info(module_type, module_name)
        return info
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{module_type}/{module_name:path}/payloads")
async def get_compatible_payloads(module_type: str, module_name: str, user=Depends(get_current_active_user)):
    """Get compatible payloads for an exploit module."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    if module_type != "exploit":
        raise HTTPException(status_code=400, detail="Compatible payloads only available for exploit modules")

    try:
        payloads = await msf_client.get_compatible_payloads(module_name)
        return {"payloads": payloads, "count": len(payloads)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{module_type}/{module_name:path}/execute")
async def execute_module(
    module_type: str,
    module_name: str,
    request: ModuleExecuteRequest,
    user=Depends(get_current_active_user)
):
    """Execute a module with given options."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    valid_types = ["exploit", "auxiliary", "post"]
    if module_type not in valid_types:
        raise HTTPException(status_code=400, detail=f"Only {valid_types} modules can be executed via this endpoint")

    try:
        if module_type == "exploit":
            result = await msf_client.run_exploit(
                module_name,
                request.options,
                request.payload,
                request.payload_options
            )
        else:
            # For auxiliary and post modules
            result = await msf_client.run_module(
                module_type,
                module_name,
                request.options
            )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/jobs")
async def list_module_jobs(user=Depends(get_current_active_user)):
    """List all running module jobs."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        jobs = await msf_client.list_jobs()
        return {"jobs": jobs, "count": len(jobs)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/jobs/{job_id}")
async def get_module_job_info(job_id: str, user=Depends(get_current_active_user)):
    """Get information about a specific job."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        job_info = await msf_client.get_job_info(job_id)
        if not job_info:
            raise HTTPException(status_code=404, detail="Job not found")
        return job_info
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: backend_ruby/standalone/routes/modules.rb
# frozen_string_literal: true

require 'sinatra/base'

module MsfGui
  module Routes
    class Modules < Sinatra::Base
      helpers Helpers::Response
      helpers Helpers::Validators

      configure do
        set :show_exceptions, false
      end

      VALID_TYPES = %w[exploit payload auxiliary post encoder nop evasion].freeze

      # Get available module types
      get '/types' do
        json_response(VALID_TYPES)
      end

      # Get module statistics (public endpoint)
      get '/stats' do
        begin
          stats = MsfGui.msf_client.get_stats
          json_response(stats)
        rescue StandardError => e
          json_response({
            exploit: 0,
            auxiliary: 0,
            post: 0,
            payload: 0,
            encoder: 0,
            nop: 0,
            error: e.message
          })
        end
      end

      # Search modules
      get '/search' do
        query = params[:query] || params[:q] || ''
        type = params[:type]

        if query.empty?
          json_error('Query parameter is required', status: 422)
        end

        modules = MsfGui.msf_client.search_modules(query, type: type)

        # Format results
        result = modules.map do |mod|
          {
            name: mod['name'] || mod['fullname'],
            fullname: mod['fullname'],
            type: mod['type'],
            rank: mod['rank'],
            description: mod['description'],
            author: mod['author'],
            references: mod['references']
          }
        end

        json_response(result)
      end

      # List modules by type with pagination
      get '/:module_type' do
        module_type = params[:module_type]

        unless VALID_TYPES.include?(module_type)
          json_error("Invalid module type: #{module_type}", status: 400)
        end

        offset = (params[:offset] || 0).to_i
        limit = (params[:limit] || 100).to_i
        search = params[:search]

        modules = MsfGui.msf_client.list_modules(module_type)

        # Apply search filter if provided
        if search && !search.empty?
          search_lower = search.downcase
          modules = modules.select { |m| m.downcase.include?(search_lower) }
        end

        # Paginate
        total = modules.size
        modules = modules[offset, limit] || []

        json_response({
          modules: modules,
          total: total,
          offset: offset,
          limit: limit
        })
      end

      # Get module info
      get '/:module_type/:module_name/info' do
        module_type = params[:module_type]
        module_name = params[:module_name]

        # Handle nested module names (e.g., windows/smb/ms17_010_eternalblue)
        if params[:splat] && !params[:splat].empty?
          module_name = "#{module_name}/#{params[:splat].join('/')}"
        end

        # URL decode the module name
        module_name = URI.decode_www_form_component(module_name) if module_name.include?('%')

        unless VALID_TYPES.include?(module_type)
          json_error("Invalid module type: #{module_type}", status: 400)
        end

        info = MsfGui.msf_client.get_module_info(module_type, module_name)

        if info['error']
          json_error(info['error_message'] || 'Module not found', status: 404)
        end

        json_response({
          name: info['name'],
          fullname: info['fullname'],
          type: module_type,
          rank: info['rank'],
          description: info['description'],
          author: info['author'],
          license: info['license'],
          references: info['references'],
          options: info['options'],
          targets: info['targets'],
          default_target: info['default_target'],
          privileged: info['privileged'],
          platform: info['platform'],
          arch: info['arch'],
          disclosure_date: info['disclosure_date']
        })
      end

      # Get compatible payloads for an exploit
      get '/:module_type/:module_name/payloads' do
        module_type = params[:module_type]
        module_name = params[:module_name]

        if params[:splat] && !params[:splat].empty?
          module_name = "#{module_name}/#{params[:splat].join('/')}"
        end

        module_name = URI.decode_www_form_component(module_name) if module_name.include?('%')

        unless module_type == 'exploit'
          json_error('Payloads are only available for exploit modules', status: 400)
        end

        payloads = MsfGui.msf_client.get_compatible_payloads(module_name)
        json_response(payloads)
      end

      # Execute module
      post '/:module_type/:module_name/execute' do
        module_type = params[:module_type]
        module_name = params[:module_name]

        if params[:splat] && !params[:splat].empty?
          module_name = "#{module_name}/#{params[:splat].join('/')}"
        end

        module_name = URI.decode_www_form_component(module_name) if module_name.include?('%')

        unless VALID_TYPES.include?(module_type)
          json_error("Invalid module type: #{module_type}", status: 400)
        end

        body = parse_json_body
        options = body[:options] || body['options'] || {}
        payload = body[:payload] || body['payload']
        payload_options = body[:payload_options] || body['payload_options'] || {}

        result = if module_type == 'exploit' && payload
                   MsfGui.msf_client.run_exploit(
                     module_name,
                     options,
                     payload: payload,
                     payload_options: payload_options
                   )
                 else
                   MsfGui.msf_client.run_module(module_type, module_name, options)
                 end

        if result['error']
          json_error(result['error_message'] || 'Module execution failed', status: 500)
        end

        json_response({
          success: true,
          job_id: result['job_id'],
          uuid: result['uuid'],
          message: "#{module_type.capitalize} module started"
        })
      end

      # List running jobs
      get '/jobs' do
        jobs = MsfGui.msf_client.list_jobs

        result = jobs.map do |id, name|
          { id: id, name: name }
        end

        json_response(result)
      end

      # Get job info
      get '/jobs/:job_id' do
        info = MsfGui.msf_client.get_job_info(params[:job_id])

        if info.nil? || info.empty?
          json_error('Job not found', status: 404)
        end

        json_response(info)
      end

      # Kill job
      delete '/jobs/:job_id' do
        result = MsfGui.msf_client.kill_job(params[:job_id])
        json_response({
          success: true,
          message: 'Job terminated',
          result: result
        })
      end
    end
  end
end
